text/0000-unsized-rvalues.md

```diff
@@ -0,0 +1,203 @@
+- Feature Name: unsized_locals
+- Start Date: 2017-02-11
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow for local variables, function arguments, and some expressions to have an unsized type, and implement it by storing the temporaries in variably-sized allocas.
+
+Have repeat expressions with a length that captures local variables be such an expression, returning an `[T]` slice.
+
+Provide some optimization guarantees that unnecessary temporaries will not create unnecessary allocas.
+
+# Motivation
+[motivation]: #motivation
+
+There are 2 motivations for this RFC:
+
+1. Passing unsized values, such as trait objects, to functions by value is often desired. Currently, this must be done through a `Box<T>` with an unnecessary allocation.
+
+  One particularly common example is passing closures that consume their environment without using monomorphization. One would like for this code to work:
+
+  ```Rust
+  fn takes_closure(f: FnOnce()) { f(); }
```

Comment:
@burdges In that scenario, is the trait/unsized type required to be object-safe? Also, for back-compat reasons, this only works at the top-level, right? I can't have `fn takes_opt_closure(f: Option<FnOnce()>)` for example, because that would overlap with trait object syntax.